0027-tcp-load-balancer.md

```diff
@@ -0,0 +1,264 @@
+---
+authors: Chris Coulter (c.coulter@gmail.com)
+state: draft
+---
+
+# RFD 27 - TCP Load Balancer
+
+## What
+
+A Toy TCP Load Balancer utilising mTLS to authenticate clients and distribute load.
+
+## Why
+
+Coding challenges can give a reasonable expectation of my performance *if* employed by Teleport.
+
+## Details
+
+### Assumptions
+
+- There is only one type of upstream server. This could be changed if using Application Layer Protocol Negotiation extension as part of TLS.
+- Upstream server connection will not include any authntication/authorisation and is a straight tcp connection.
+- Upstream server is expected to implement ping/pong functionality.
+- Upstream servers are grouped and tcp addresses / ports will be hard coded.
+- Authorisation for a client to a server group will be hard coded.
+- The CA, Server keys + cert and client keys + cert will be precomputed using openssl for creation and signing.
+- Client certificate public keys will be used to for authorization data.
+- Maxmium connection count per client will be hard coded to 10.
+
+### Library
+- Single threaded
+    - simplicity
+    - much harder to deadlock
+    - at expense of throughput
+    - could offload to threads or use async
+    - at expense of possible deadlocks, greater complexity.
+
+
+```rust
+struct LoadBalancer
+{
+    clients         : HashMap<String, Client>,
+    server_groups   : HashMap<u32, ServerGroup>,
```

Comment:
What would prevent you from allocating consecutive integers starting from 0 to your `ServerGroup`s?